pkg/apiclient/headless/headless.go

```diff
@@ -0,0 +1,243 @@
+package headless
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/alicebob/miniredis/v2"
+	"github.com/go-redis/redis/v8"
+	"github.com/golang/protobuf/ptypes/empty"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/pflag"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	cache2 "k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/utils/pointer"
+
+	"github.com/argoproj/argo-cd/v2/pkg/apiclient"
+	"github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	appclientset "github.com/argoproj/argo-cd/v2/pkg/client/clientset/versioned"
+	repoapiclient "github.com/argoproj/argo-cd/v2/reposerver/apiclient"
+	"github.com/argoproj/argo-cd/v2/server"
+	servercache "github.com/argoproj/argo-cd/v2/server/cache"
+	"github.com/argoproj/argo-cd/v2/util/cache"
+	appstatecache "github.com/argoproj/argo-cd/v2/util/cache/appstate"
+	"github.com/argoproj/argo-cd/v2/util/cli"
+	"github.com/argoproj/argo-cd/v2/util/io"
+	kubeutil "github.com/argoproj/argo-cd/v2/util/kube"
+	"github.com/argoproj/argo-cd/v2/util/localconfig"
```

Comment:
I think we might be bringing in too many dependencies into `pkg/apiclient`, which is meant to be a library/SDK. I think this package still belongs next to CLI code rather than library code.

Reply:
If we are okay with shifting the `NewClientOrDie()` to the CLI (which should be okay because it's only used there) we can move this pkg to the cli part. NewClientOrDie needs to access the headless methods which if kept in the apiClient pkg separate from headless in cli, causes import cycles.

Reply:
> If we are okay with shifting the NewClientOrDie() to the CLI (which should be okay because it's only used there) we can move this pkg to the cli part. 

Spoke with @alexmt and he agrees that we should move this logic into CLI specific package to keep dependencies of pkg/apiclient leaner.

Reply:
Will shift it to CLI

Reply:
fixed